### PR TITLE
Add working directory sensitivity for `restore` and `deploy`

### PR DESCRIPTION
**Pull request by weikanglim**
Allow `deploy` and `restore` to observe working directory to deploy all services, or a particular service.

1. `azd deploy` and `azd restore` will now deploy the current service, when the current working directory is set to a service directory. 
2. `azd deploy` and `azd restore` will deploy all services, when the current working directory is set to the root project directory. 
3. In other directories, `azd deploy` and `azd restore` will not attempt a deployment and instead offer suggestions. `azd deploy --all` can be used to deploy all services, or `azd deploy <service>` to deploy a given service. 

End-to-end tests have been added for `azd restore` and `azd deploy` that validates the working and error behavior.

Fixes https://github.com/Azure/azure-dev-pr/issues/1498

### CODE REVIEW

File: cli/azd/pkg/project/project_config.go [hunk 1]
```diff
@@ -54,3 +55,19 @@ func (p *ProjectConfig) HasService(name string) bool {
 
 	return false
 }
+
+// Retrieves the list of services in the project, in a stable ordering that is deterministic.
+func (p *ProjectConfig) GetServicesStable() []*ServiceConfig {
```
Comment (weikanglim) [inline cli/azd/pkg/project/project_config.go:60]: We will likely revisit this again shortly, but just wanted to restore the behavior and avoid the regression as described by #1816

**Review by ellismg — APPROVED** [review]
LGTM

File: cli/azd/cmd/deploy.go [hunk 2]
```diff
@@ -150,22 +160,47 @@ func (d *deployAction) Run(ctx context.Context) (*actions.ActionResult, error) {
...
 	}
 
+	if d.flags.all && targetServiceName != "" {
+		return nil, fmt.Errorf("cannot specify both --all and <service>")
+	}
+
+	if !d.flags.all && targetServiceName == "" {
+		var err error
+		targetServiceName, err = defaultServiceFromWd(d.azdCtx, d.projectConfig)
+		if err == errNoDefaultService {
```
[the remark this answers is not in the document]
Comment (weikanglim) [reply cli/azd/cmd/deploy.go:170]: 👌looks like our linter caught this one. Will change.

**Comment by weikanglim** [comment]
Overriding check-enforcer. Installer tests are failing that are unrelated to the change.

**Pull request merged**